drivers/gpio/gpio_esp32.c

```diff
@@ -137,37 +106,44 @@ static int gpio_esp32_config(struct device *dev, int access_op,
 	struct gpio_esp32_data *data = dev->driver_data;
 	u32_t func;
 	int r;
+	u32_t io_pin = pin + data->port.pin_offset; /* Range from 0 - 39 */
 
 	if (access_op != GPIO_ACCESS_BY_PIN) {
 		return -ENOTSUP;
 	}
 
 	/* Query pinmux to validate pin number. */
-	r = pinmux_pin_get(data->pinmux, pin, &func);
+	r = pinmux_pin_get(data->pinmux, io_pin, &func);
 	if (r < 0) {
 		return r;
 	}
 
-	pinmux_pin_set(data->pinmux, pin, PIN_FUNC_GPIO);
-	if (flags & GPIO_PUD_PULL_UP) {
-		pinmux_pin_pullup(data->pinmux, pin, PINMUX_PULLUP_ENABLE);
-	} else if (flags & GPIO_PUD_PULL_DOWN) {
-		pinmux_pin_pullup(data->pinmux, pin, PINMUX_PULLUP_DISABLE);
+	pinmux_pin_set(data->pinmux, io_pin, PIN_FUNC_GPIO);
+	if (flags & GPIO_PULL_UP) {
+		pinmux_pin_pullup(data->pinmux, io_pin, PINMUX_PULLUP_ENABLE);
+	} else if (flags & GPIO_PULL_DOWN) {
+		pinmux_pin_pullup(data->pinmux, io_pin, PINMUX_PULLUP_DISABLE);
 	}
 
-	if (flags & GPIO_DIR_OUT) {
-		r = pinmux_pin_input_enable(data->pinmux, pin,
-					PINMUX_OUTPUT_ENABLED);
+	if (flags & GPIO_OUTPUT) {
+		r = pinmux_pin_input_enable(data->pinmux, io_pin,
+					    PINMUX_OUTPUT_ENABLED);
 		assert(r >= 0);
 	} else {
-		pinmux_pin_input_enable(data->pinmux, pin,
+		pinmux_pin_input_enable(data->pinmux, io_pin,
 					PINMUX_INPUT_ENABLED);
-		config_polarity(pin, flags);
+		config_polarity(io_pin, flags);
 	}
 
-	config_drive_strength(pin, flags);
+	config_drive_strength(io_pin, flags);
+
+	if (flags & GPIO_OUTPUT_INIT_HIGH) {
+		*data->port.write.set_reg = BIT(pin);
+	} else if (flags & GPIO_OUTPUT_INIT_LOW) {
+		*data->port.write.clear_reg = BIT(pin);
+	}
```

Comment:
Initial pin value should be set before we configure the pin in the output mode. Otherwise we may see a glitch.